### DISCUSSION TITLE
adding fields

### DISCUSSION THREAD

**Discussion by ghost** [dtcenter/METplus#2518] [Configuration] [answered]
Hi,
I am adding two surface concentration fields using pcpcombine step in Metplus from a grib file and found it is only taking one field, instead of two in the output. Is there something needed to be set up differently in config file. I have attached the config file and logfile in here for reference. Any help is appreciated.
Thanks
[metplus.log.20240320193032.txt](https://github.com/dtcenter/METplus/files/14672003/metplus.log.20240320193032.txt)
[Pointstat_AirnowPM25_comb.conf.txt](https://github.com/dtcenter/METplus/files/14672008/Pointstat_AirnowPM25_comb.conf.txt)

**Comment by CPKalb** [comment, marked as answer]
Hi:

The easiest way to accomplish this addition would be to use the [User-Defined](https://metplus.readthedocs.io/en/latest/Users_Guide/wrappers.html#user-defined-commands) option in METplus.  For a User-Defined command, the exact command to be run in PCP-Combine is specified directly.  

To run the addition this way, you would want to set the following in your configuration file:
FCST_PCP_COMBINE_METHOD = USER_DEFINED
FCST_PCP_COMBINE_COMMAND = -add {FCST_PCP_COMBINE_INPUT_DIR}/{FCST_PCP_COMBINE_INPUT_TEMPLATE} 'name="MASSDEN"; level="R807";' FCST_PCP_COMBINE_INPUT_DIR}/{FCST_PCP_COMBINE_INPUT_TEMPLATE} 'name="MASSDEN"; level="R808";' 

You should also be able to remove FCST_PCP_COMBINE_LOOKBACK, FCST_PCP_COMBINE_CONSTANT_INIT, FCST_PCP_COMBINE_INPUT_ACCUMS, FCST_PCP_COMBINE_INPUT_NAMES, and FCST_PCP_COMBINE_INPUT_LEVELS.

The command listed in the resulting METplus log should show both variables, and also show that it is reading both variables.  The MET output would look something like the following:
DEBUG 2: Since the "-field" command line option was not used, parsing the command line arguments a list of files, each followed by a configuration string.
DEBUG 2: Performing derivation command (sum) for 2 files.
DEBUG 1: Reading data (name="MASSDEN"; level="R807";) from input file: /lfs/h2/emc/ptmp/emc.lam/rrfs/v0.8.5/prod/rrfs.20240319/00/rrfs.t00z.prslev.f000.conus_3km.grib2
DEBUG 1: Reading data (name="MASSDEN"; level="R808";) from input file: /lfs/h2/emc/ptmp/emc.lam/rrfs/v0.8.5/prod/rrfs.20240319/00/rrfs.t00z.prslev.f000.conus_3km.grib2
DEBUG 1: Creating output file: /lfs/h2/emc/physics/noscrub/partha.bhattacharjee/RRFSSD_Eval/Airnow/pcpcomb/20240319/pm25tot_2024031900_000.nc
DEBUG 2: Writing output variable "PMTF" for the "sum" of "MASSDEN/R807".

Does this produce the expected output?

Christina

**Reply by ghost** [reply]
Thanks @CPKalb. I have modified config accordingly and here is the logfile. 
[metplus.log.20240323150209.txt.txt](https://github.com/dtcenter/METplus/files/14747824/metplus.log.20240323150209.txt.txt)
[PcpCombine_AirnowPM25.conf.txt.txt](https://github.com/dtcenter/METplus/files/14747826/PcpCombine_AirnowPM25.conf.txt.txt)

**Reply by CPKalb** [reply]
From the log file, this looks like it is working as expected.  Does the output data look as expected?

Christina

**Reply by ghost** [reply]
I am looking into it, as input data were in Lambert conformal grid, whether that is retained in final netcdf file lat/lon need to be checked.

**Reply by CPKalb** [reply]
The grid information as output from MET should be given in the global attributes of the netCDF file.  If it's in Lambert Conformal, it should state this in the Projection global attribute.

Christina

**Comment by CPKalb** [comment]
Hi:

I am checking in on this.  Is it working as expected or are there still modifications needed?

Thanks,
Christina

**Reply by ghost** [reply]
Apologies for late reply. I looked into the result and seems to be adding two fields did not work as expected. I have added two grib2 fields that are used in pcpcombine as well as plotting the added field from plotting script. Last plot on bottom right uses netcdf field generated from pcpcombine and it differs considerably from what is expected.
![PM25_DustSmokeGrbNetcdf 20240202F00](https://github.com/dtcenter/METplus/assets/88152718/bee2e3c9-5d41-48da-b860-a584d0b30491)

**Reply by CPKalb** [reply]
Hi:

To further debug this, I will need to your METplus configuration file as well as the two netCDF files that you are trying to add.  Instructions on how to send us data can be found [here](https://github.com/dtcenter/METplus/discussions/954).

Thanks,
Christina

**Reply by ghost** [reply]
Thanks Christina. I have uploaded netcdf file, input grib file, configue file,  in ftp under ftp://anonymous@ftp.rap.ucar.edu:21/incoming/irap/met_help/perthsb 
Let me know if you need anything else

**Comment by CPKalb** [comment]
Hi:

When I look in the data file sent (rrfs.t00z.prslev.f000.conus_3km.grib2), the field MASSDEN is not record number 807 or 808.  In this file, MASSDEN is the variable name for records 790, 791, and 792.  Assuming you want to perform the addition on the variable MASSDEN, MET is not adding the correct fields with this configuration.  I'm guessing that perhaps you want to add  levels 790 and 791.  If I switch the MET configuration levels to be R790 and R791, then the I get the following:

Added output :
![test_output](https://github.com/dtcenter/METplus/assets/22180338/a1a2f8fb-e75f-4639-bc57-dd599f11e519)


Input field 790:
![test_input1](https://github.com/dtcenter/METplus/assets/22180338/3a0ccc24-e02a-43e9-8e8d-afcd6829bb5a)


Input field 791:
![test_input2](https://github.com/dtcenter/METplus/assets/22180338/0a8ee0a6-3c8d-4226-96c2-a56c36e9794f)

The color bar ranges are not exactly the same, but it appears to be adding as expected.

Using record numbers as the level can be tricky and is generally not recommended since record numbers can change between files, giving inconsistent results.  There are other grib1/2 options available in MET to help select the correct field when variable names are duplicated.  They are listed in the [fcst](https://met.readthedocs.io/en/latest/Users_Guide/config_options.html#fcst) section of the MET documentation (you will have to scroll down a little to get to the grib1/2 options).  In this case, you can set GRIB2_ipdtmpl_index and GRIB2_ipdtmpl_val to help distinguish between fields instead of using record numbers.   For the file you sent, it would look like this for the different MASSDEN variables:

Record 790:
'name="MASSDEN"; level="L0"; GRIB2_ipdtmpl_index=2; GRIB2_ipdtmpl_val=62010;'

Record 791:
'name="MASSDEN"; level="L0"; GRIB2_ipdtmpl_index=[2,6]; GRIB2_ipdtmpl_val=[62001,0];'

Record 792:
'name="MASSDEN"; level="L0"; GRIB2_ipdtmpl_index=6; GRIB2_ipdtmpl_val=7;'

Assuming you want to add records 790 and 791, the FCST_PCP_COMBINE_COMMAND would need to be changed to look like this:

FCST_PCP_COMBINE_COMMAND = -add {FCST_PCP_COMBINE_INPUT_DIR}/{FCST_PCP_COMBINE_INPUT_TEMPLATE} 'name="MASSDEN"; level="L0"; GRIB2_ipdtmpl_index=2; GRIB2_ipdtmpl_val=62010;' {FCST_PCP_COMBINE_INPUT_DIR}/{FCST_PCP_COMBINE_INPUT_TEMPLATE} 'name="MASSDEN"; level="L0"; GRIB2_ipdtmpl_index=[2,6]; GRIB2_ipdtmpl_val=[62001,0];'

Changing the command this way produces the same added output as the image above without needing to rely on record numbers.  Does this produce the expected results on your end?

Christina

**Reply by CPKalb** [reply]
Hi:

I'm checking back in on this issue.  Did the above produce the expected output?  If so, I would like to close out the discussion.

Thanks,
Christina

**Reply by ghost** [reply]
Sorry for late reply, I have tested it and it is working fine. Thanks for your help.

**Reply by CPKalb** [reply]
Great, I'm going to close and lock the discussion to prevent future posts.  If you have additional questions, feel free to start a new discussion.

Christina
